### ISSUE TITLE
Add SOVERSION to shared lib to follow linux conventions

### ISSUE THREAD

**Issue by ykoehler** [sysrepo/sysrepo#788]
The shared libraries shared by this package are not following Linux convention and only output a lib*.so file.  The SOVERSION target properties should be specified so that the proper lib*.so.<maj>, lib*.so.<maj>.<min> and lib*.so.<maj>.<min>.<patch> symlinks gets generated.

**Comment by rkrejci** [comment]
@mislavn - is it reasonable to add also soname version into the binding libraries created by SWIG? I've added it only into the main libsysrepo.so.

**Comment by mislavn** [comment]
> @mislavn - is it reasonable to add also soname version into the binding libraries created by SWIG? I've added it only into the main libsysrepo.so.

You are right, I will make a pull request.

**Comment by rkrejci** [comment]
@ykoehler, could you please describe your usecase to have soname set in libsysrepo library? I mean, do you have some issue with not having soname or is it just "follow common convention"? The libsysrepo is used to communicate with the main sysrepod daemon, so it may be a bad idea to allow having multiple libsysrepo version in use and communicate with sysrepod of a different version.

**Comment by ykoehler** [comment]
The issue I have with the lib missing soname is that I use this project inside Yocto, and Yocto will "split" installed files into multiple .rpm packages, and it doesn't recognise the .so file unless it follow the convention.  The file therefore is not part of the main *.rpm, and get move to the *-devel.rpm.  With the SONAME, the lib is then automatically put in the main *.rpm as expected.

**Comment by rkrejci** [comment]
Ok, proposed solution:
- use soname, so app always links the correct library version. We will get this:
```
libsysrepo.so -> libsysrepo.so.0.5.11
libsysrepo.so.0.5 -> libsysrepo.so.0.5.11
libsysrepo.so.0.5.11
```
- send version number in `sr_connect()` (client side) and verify it in somewhere in the connection manager (in sysrepod process) to have the same (soname) version on both sides.

Another approach is to avoid SONAME in libsysrepo (Yocto as well as any other RPM builder should be then forced to put the .so into the correct package) and having instead a check for build-time and run-time libsysrepo version compatibility (version check in `sr_connect()` is necessary anyway). But the approach with SONAME seems more generic and standard.

Any thoughts or comments? Did I miss something?

**Comment by rkrejci** [comment]
small correction, symlinks are actually
```
libsysrepo.so -> libsysrepo.so.0.5
libsysrepo.so.0.5 -> libsysrepo.so.0.5.11
libsysrepo.so.0.5.11
```

**Comment by ykoehler** [comment]
I also had to alter the libsysrepo-cpp to do the same trick.

**Comment by ykoehler** [comment]
From a5507614041675e82a106a1a367ed15421abb552 Mon Sep 17 00:00:00 2001
From: Yannick Koehler <yannick.koehler@hpe.com>
Date: Tue, 11 Apr 2017 15:17:58 -0400
Subject: [PATCH] Added version to .so file

---
 src/CMakeLists.txt  |    7 +++++--
 swig/CMakeLists.txt |    5 +++++
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 54e963e..8d79a0f 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -99,10 +99,13 @@ endif()
 
 # sysrepo library
 add_library(sysrepo SHARED $<TARGET_OBJECTS:COMMON> $<TARGET_OBJECTS:SR_ENGINE> $<TARGET_OBJECTS:SR_SRC>)
+SET_TARGET_PROPERTIES(sysrepo PROPERTIES
+              OUTPUT_NAME sysrepo VERSION 1 SOVERSION 1.0)
+target_link_libraries(sysrepo ${LINK_LIBRARIES})
+
 add_library(sysrepo_a $<TARGET_OBJECTS:COMMON> $<TARGET_OBJECTS:SR_ENGINE> $<TARGET_OBJECTS:SR_SRC>)
 SET_TARGET_PROPERTIES(sysrepo_a PROPERTIES
-              OUTPUT_NAME sysrepo CLEAN_DIRECT_OUTPUT 1)
-target_link_libraries(sysrepo ${LINK_LIBRARIES})
+              OUTPUT_NAME sysrepo CLEAN_DIRECT_OUTPUT 1 VERSION 1 SOVERSION 1.0)
 target_link_libraries(sysrepo_a ${LINK_LIBRARIES})
 
 # sysrepo daemon
diff --git a/swig/CMakeLists.txt b/swig/CMakeLists.txt
index 3a6483a..00082c5 100644
--- a/swig/CMakeLists.txt
+++ b/swig/CMakeLists.txt
@@ -86,6 +86,11 @@ set(SYSREPO_H_SOURCES
 	${CMAKE_SOURCE_DIR}/swig/cpp/src/Internal.h)
 
 add_library(Sysrepo-cpp SHARED ${SYSREPO_CPP_SOURCES})
+set_target_properties(Sysrepo-cpp PROPERTIES
+  VERSION 1
+  SOVERSION 1.0
+  OUTPUT_NAME Sysrepo-cpp
+  )
 
 # include custom Modules
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/CMakeModules/")
-- 
1.7.1

**Comment by ykoehler** [comment]
here is my patch file.

**Comment by ykoehler** [comment]
[0002-Added-version-to-.so-file.patch.zip](https://github.com/sysrepo/sysrepo/files/964445/0002-Added-version-to-.so-file.patch.zip)

**Comment by mislavn** [comment]
> I also had to alter the libsysrepo-cpp to do the same trick.

When we agree on the versioning I will alter the so files of the language bindings, that's not a problem.

**Comment by rkrejci** [comment]
regarding the patch - I believe that `VERSION` and `SOVERSION` properties in case of static libsysrepo.a are useless, aren't they?

**Comment by mislavn** [comment]
> regarding the patch - I believe that VERSION and SOVERSION properties in case of static libsysrepo.a are useless, aren't they?

Because the sysrepo static binary is not self contained I think it would be misleading to add `VERSION` and `SOVERSION` to it.

**Comment by rkrejci** [comment]
fixed by #833
